### PR TITLE
Perform "mini scheduling run" after task has finished

### LINKED ISSUE #11498: Perform "mini scheduling run" after task has finished
In order to further reduce intra-dag task scheduling lag we can add a further optimization: when a task has just finshed executing (success or failure) we can look at the downstream tasks of the task that just finished, and then make scheduling decisions for those tasks there -- we've already got the dag loaded, and we know they are _likely_ actionable as we just finished. We should set tasks to scheduled if we can (but no further, i.e. not to queued).

Upstream_failed may also not be safe as that would run callbacks. Think about this.

We'll need a config option to disable this, as it _could_ lead to task starvation of other dags

### PR DESCRIPTION
**Pull request by jhtimmins**
closes https://github.com/apache/airflow/issues/11498

In order to further reduce intra-dag task scheduling lag we add an
optimization: when a task has just finished executing (success or
failure) we can look at the downstream tasks of just that task, and then
make scheduling decisions for those tasks there -- we've already got the
dag loaded, and we know they are likely actionable as we just finished.

We should set tasks to scheduled if we can (but no further, i.e. not to
queued, as the scheduler has to make that decision with info about the
Pool usage etc.).

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -1664,42 +1664,11 @@ def _schedule_dag_run(self, dag_run: DagRun, currently_active_runs: int, session
...
-            ti for ti in schedulable_tis
-            if
-            (
-                ti.task.task_type == "DummyOperator"
-                and not ti.task.on_execute_callback
-                and not ti.task.on_success_callback
-            )
-        ]
-
         # This will do one query per dag run. We "could" build up a complex
```
Comment (mik-laj) [inline airflow/jobs/scheduler_job.py:1655]: It seems to me that we should also move the comment. Now it has lost context.

File: airflow/models/dagrun.py [hunk 2]
```diff
@@ -43,6 +44,20 @@
 from airflow.utils.types import DagRunType
 
 
+class _TISchedulingDecision(TypedDict):
```
Comment (mik-laj) [inline airflow/models/dagrun.py]: Is there a reason we need to use TypedDict? NamedTuple is much easier to use in many cases.

File: airflow/jobs/scheduler_job.py (hunk 1 above)
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1655]: This comment doesn't make sense when called on a (instance) method on DagRun, as that almost by definition only operators on a single dag run. The comment is kept here in the scheduler because that's where might think we want to batch the queries up, but shouldn't.

File: airflow/models/dagrun.py (hunk 2 above)
Comment (ashb) [reply airflow/models/dagrun.py]: No reason -- just happened to be the example we were looking at in PoolStats and copied that.

File: airflow/models/dagrun.py [hunk 3]
```diff
@@ -466,7 +475,35 @@ def update_state(
...
+        schedulable_tis: List[TI] = []
+        changed_tis = False
+
+        tis = list(self.get_task_instances(session=session, state=State.task_states + (State.SHUTDOWN,)))
+        self.log.debug("number of tis tasks for %s: %s task(s)", self, len(tis))
+        for ti in tis:
+            ti.task = self.get_dag().get_task(ti.task_id)
+
+        unfinished_tasks = [t for t in tis if t.state in State.unfinished]
+        finished_tasks = [t for t in tis if t.state in State.finished | {State.UPSTREAM_FAILED}]
```
Comment (ashb) [inline airflow/models/dagrun.py]:
```suggestion
        finished_tasks = [t for t in tis if t.state in State.finished]
```

I added UPSTREAM_FAILED in to `finished` in a PR that's already been merged.

File: airflow/models/dagrun.py [hunk 4]
```diff
@@ -380,27 +395,21 @@ def update_state(
...
-                self, len(scheduleable_tasks))
-            ready_tis, changed_tis = self._get_ready_tis(scheduleable_tasks, finished_tasks, session)
-            self.log.debug("ready tis length for %s: %s task(s)", self, len(ready_tis))
-            if none_depends_on_past and none_task_concurrency:
-                # small speed up
-                are_runnable_tasks = ready_tis or self._are_premature_tis(
-                    unfinished_tasks, finished_tasks, session) or changed_tis
+
+        if unfinished_tasks and none_depends_on_past and none_task_concurrency:
+            # small speed up
```
Comment (jhtimmins) [inline airflow/models/dagrun.py:405]: @ashb It looks like `are_runnable_tasks` isn't used until line 456. Is there a reason this conditional needs to be here rather than right before it's used?

Comment (ashb) [reply airflow/models/dagrun.py:405]: Bad/lazy refactor most likely.

Comment (jhtimmins) [reply airflow/models/dagrun.py:405]: Ah looks like it's there so it's outside the conditional. nvm.

File: tests/models/test_dagrun.py [hunk 5]
```diff
@@ -215,6 +218,110 @@ def test_dagrun_success_conditions(self):
...
+            task_a >> task_b
+            task_b >> task_c
+
+        scheduler = SchedulerJob()
+        scheduler.dagbag.bag_dag(dag, root_dag=dag)
+
+        dag_run = dag.create_dagrun(run_id='test_dagrun_fast_follow', state=State.RUNNING)
+
+        task_instance_a = dag_run.get_task_instance(task_id=task_a.task_id)
+        task_instance_a.task = dag.get_task(task_a.task_id)
```
Comment (ashb) [inline tests/models/test_dagrun.py]:
```suggestion
        task_instance_a.task = task_a
```

File: tests/models/test_dagrun.py [hunk 6]
```diff
@@ -215,6 +218,110 @@ def test_dagrun_success_conditions(self):
...
+
+        task_instance_b = dag_run.get_task_instance(task_id=task_b.task_id)
+        task_instance_b.task = dag.get_task(task_b.task_id)
```
Comment (ashb) [inline tests/models/test_dagrun.py]:
```suggestion
        task_instance_b.task = task_b
```

File: tests/models/test_dagrun.py [hunk 7]
```diff
@@ -215,6 +218,110 @@ def test_dagrun_success_conditions(self):
...
+        schedulable_tis, _ = dag_run.update_state()
+        self.assertEqual(len(schedulable_tis), 1)
+        self.assertEqual(schedulable_tis[0].task_id, task_a.task_id)
+        self.assertEqual(schedulable_tis[0].state, State.NONE)
+
+        dag_run.schedule_tis(schedulable_tis, session)
+        self.validate_ti_states(dag_run, {'A': State.SCHEDULED, 'B': State.NONE, 'C': State.NONE})
+
+        scheduler._critical_section_execute_task_instances(session=session)
+        self.validate_ti_states(dag_run, {'A': State.QUEUED, 'B': State.NONE, 'C': State.NONE})
```
Comment (ashb) [inline tests/models/test_dagrun.py]: Do we need this block? I think this test would be clearer if we instead just directly set task_instance_a to a runnable state: For example:


```python
        task_instance_a.state = State.QUEUED
        session.commit()
```

My reason here is that this is the "pre-condition/setup" for the test, not part of what are actually testing here, so by having these asserts and calling the scheduler job code we are not testing this feature in isolation.

Comment (ashb) [reply tests/models/test_dagrun.py]: (For this to work the TI would need to be attached to the session you would need to pass `session=session` to dag_run.get_task_instance)

**Review by ashb — CHANGES_REQUESTED** [review]
We can also possibly reduce the duplication in tests between `test_dagrun_fast_follow` and `test_dagrun_fast_follow_deactiveated` by using `@parameterized.expand` and `with conf_vars(...):`

Oh also, these new tests should probably be in tests/models/test_taskinstance.py -- as the code we are testing is in TaskInstance primarily.

**Comment by kaxil** [comment]
Whoops, alot of tests are failing: https://github.com/apache/airflow/pull/11589/checks?check_run_id=1330520635

File: airflow/models/dag.py [hunk 8]
```diff
@@ -1394,36 +1402,54 @@ def sub_dag(self, *args, **kwargs):
...
+
         also_include = []
-        for t in regex_match:
+        for t in matched_tasks:
             if include_downstream:
                 also_include += t.get_flat_relatives(upstream=False)
             if include_upstream:
                 also_include += t.get_flat_relatives(upstream=True)
+            if include_direct_parents and not include_upstream:
+                also_include += t.get_flat_relatives(upstream=True, recurse=False)
```
Comment (ashb) [inline airflow/models/dag.py]:
```suggestion
            elif include_direct_parents:
                also_include += t.upstream_list
```

And then we don't need to add `recurse=True` parameter to get_flat_relatives (the "flat" part is to do the flattening/recursing part)

File: airflow/models/dagrun.py [hunk 9]
```diff
@@ -43,6 +43,20 @@
 from airflow.utils.types import DagRunType
 
 
+class _TISchedulingDecision(NamedTuple):
```
Comment (ashb) [inline airflow/models/dagrun.py]:
```suggestion
class TISchedulingDecision(NamedTuple):
```

I don't know why I made this "private" -- and it looks a bit odd to use it in a return value as it was

File: tests/models/test_dagrun.py [hunk 10]
```diff
@@ -45,6 +45,7 @@ def setUpClass(cls):
     def setUp(self):
         clear_db_runs()
         clear_db_pools()
+        clear_db_dags()
```
Comment (ashb) [inline tests/models/test_dagrun.py]: Is this still needed, or can we revert the changes to this file?

File: tests/models/test_taskinstance.py [hunk 11]
```diff
@@ -1670,6 +1754,100 @@ def test_get_rendered_template_fields(self):
...
+            {'A': State.QUEUED, 'B': State.NONE, 'C': State.NONE},
+            {'A': State.SUCCESS, 'B': State.NONE, 'C': State.NONE},
+            None
+        ),
+        (
+            {('scheduler', 'schedule_after_task_execution'): 'True'},
+            {'A': 'B', 'C': 'B', 'D': 'C'},
+            {'A': State.QUEUED, 'B': State.NONE, 'C': State.NONE, 'D': State.NONE},
+            {'A': State.SUCCESS, 'B': State.NONE, 'C': State.NONE, 'D': State.NONE},
+            None
```
Comment (ashb) [inline tests/models/test_taskinstance.py]: Could you add a comment here saying what we're testing, why we expect B to not be scheduled etc -- this will help Future Us

File: airflow/models/taskinstance.py [hunk 12]
```diff
@@ -1135,7 +1136,59 @@ def _run_raw_task(
...
+                # Get a partial dag with just the specific tasks we want to
+                # examine. In order for dep checks to work correctly, we
+                # include ourself (so TriggerRuleDep can check the state of the
+                # task we just executed)
+                tasks = self.task.downstream_list + [self.task]
+                task_ids = [t.task_id for t in tasks]
+
+                partial_dag = self.task.dag.partial_subset(
+                    task_ids,
+                    include_downstream=True,
```
Comment (ashb) [inline airflow/models/taskinstance.py]: Hmmmm, this is going to include _all_ downstream tasks, recursively, not just the tasks we've selected.

**Review by ashb — APPROVED** [review]

File: airflow/models/taskinstance.py [hunk 13]
```diff
@@ -1135,7 +1136,53 @@ def _run_raw_task(
...
+                        schedulable_ti.task = self.task.dag.get_task(schedulable_ti.task_id)
+
+                num = dag_run.schedule_tis(schedulable_tis)
+                self.log.info("%d downstream tasks scheduled from follow-on schedule check", num)
+
+                session.commit()
+            except OperationalError:
+                # Any kind of DB error here is _non fatal_ as this block is just an optimisation.
+                self.log.info("DB error when checking downstream tasks ignored", exc_info=True)
+                session.rollback()
```
Comment (turbaszek) [inline airflow/models/taskinstance.py]: How about encapsulating this into separate method? The `_run_raw_task` is already a long one 😉

File: airflow/models/taskinstance.py [hunk 14]
```diff
@@ -1135,7 +1136,53 @@ def _run_raw_task(
...
+                    if not hasattr(schedulable_ti, "task"):
+                        schedulable_ti.task = self.task.dag.get_task(schedulable_ti.task_id)
+
+                num = dag_run.schedule_tis(schedulable_tis)
+                self.log.info("%d downstream tasks scheduled from follow-on schedule check", num)
+
+                session.commit()
+            except OperationalError:
+                # Any kind of DB error here is _non fatal_ as this block is just an optimisation.
+                self.log.info("DB error when checking downstream tasks ignored", exc_info=True)
```
Comment (turbaszek) [inline airflow/models/taskinstance.py]: As I user I would be worried about seeing such info logs. Should it be debug?

File: airflow/models/dagrun.py [hunk 15]
```diff
@@ -638,3 +674,52 @@ def get_latest_runs(cls, session=None):
...
+        All the TIs should belong to this DagRun, but this code is in the hot-path, this is not checked -- it
+        is the caller's responsibility to call this function only with TIs from a single dag run.
+        """
+        # Get list of TIs that do not need to executed, these are
+        # tasks using DummyOperator and without on_execute_callback / on_success_callback
+        dummy_tis = {
+            ti for ti in schedulable_tis
+            if
+            (
+                ti.task.task_type == "DummyOperator"
```
Comment (turbaszek) [inline airflow/models/dagrun.py:692]: Will it work for operators that inherit from `DummyOperator`?

File: airflow/models/dagrun.py [hunk 16]
```diff
@@ -43,6 +43,20 @@
 from airflow.utils.types import DagRunType
 
 
+class TISchedulingDecision(NamedTuple):
+    """
+    Type of return for DagRun.task_instance_scheduling_decisions
+
+    This is only used by type checkers, at run time this is a plain dict.
```
Comment (turbaszek) [inline airflow/models/dagrun.py]: Is this comment still valid?

File: airflow/models/taskinstance.py (hunk 14 above)
Comment (ashb) [reply airflow/models/taskinstance.py]: I think it depends what the exception was

File: airflow/models/dagrun.py (hunk 15 above)
Comment (ashb) [reply airflow/models/dagrun.py:692]: No, there is already an issue for that though

File: airflow/models/dagrun.py (hunk 16 above)
Comment (ashb) [reply airflow/models/dagrun.py]: No, not any more. Good catch

File: airflow/models/taskinstance.py (hunk 13 above)
Comment (jhtimmins) [reply airflow/models/taskinstance.py]: Yeah that's a good point. Will do.

File: airflow/models/dagrun.py (hunk 15 above)
Comment (ashb) [reply airflow/models/dagrun.py:692]: #11393

File: airflow/models/taskinstance.py (hunk 14 above)
Comment (turbaszek) [reply airflow/models/taskinstance.py]: It always will be a "database exception" which is rather critical one imho. And here we are telling users "your database refused something but you don't have to worry about it". I think we either have to make it less "critical" like `Skipping mini scheduling run due to exception: %s`. But still, logging the exception will show problem with database...

**Review by houqp — APPROVED** [review]

File: airflow/models/taskinstance.py (hunk 14 above)
Comment (ashb) [reply airflow/models/taskinstance.py]: The most likely case I expect here is a "cannot reach DB" network error.

But yeah, I like your message better.

File: airflow/models/taskinstance.py [hunk 17]
```diff
@@ -1135,8 +1136,63 @@ def _run_raw_task(
         if not test_mode:
             session.add(Log(self.state, self))
             session.merge(self)
+
         session.commit()
 
+        self._run_mini_scheduler_on_child_tasks(session)
```
Comment (turbaszek) [inline airflow/models/taskinstance.py:1142]: It may happened that we will do `rollback` on session we already committed, is is expected @jhtimmins ?
